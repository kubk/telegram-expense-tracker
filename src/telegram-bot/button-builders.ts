import { BankAccount, Currency } from '@prisma/client';
import { Markup } from 'telegraf';
import { BotAction, BotActionHidden, humanizeAction } from './bot-action';
import { UserTransactionExpenseRowItem } from '../repository/transaction-repository';
import { formatMoney } from './format-money';
import { currencyToSymbol } from './currency-to-symbol';

export const buildBankAccountMenu = (bankAccountId: string) => {
  return [
    [
      Markup.button.callback(
        humanizeAction(BotAction.StatisticMonths),
        `${BotAction.StatisticMonths}:${bankAccountId}`
      ),
    ],
    [
      Markup.button.callback(
        humanizeAction(BotAction.StatisticWeeks),
        `${BotAction.StatisticWeeks}:${bankAccountId}`
      ),
    ],
    [
      Markup.button.callback(
        humanizeAction(BotAction.UploadBankStatement),
        BotAction.UploadBankStatement
      ),
    ],
    [
      Markup.button.callback(
        humanizeAction(BotAction.TransactionAddManual),
        `${BotAction.TransactionAddManual}:${bankAccountId}`
      ),
    ],
    [Markup.button.callback('◀️ Back', BotAction.BankAccountList)],
  ];
};

const buildStatisticGrid = (
  row: UserTransactionExpenseRowItem,
  bankAccount: { id: string; currency: Currency }
) => {
  return [
    Markup.button.callback(
      `${row.groupname} ${formatMoney(
        row.difference,
        bankAccount.currency
      ).padStart(10, ' ')}`,
      `${BotAction.SelectStatisticsMonth}:${row.groupname}:${BotActionHidden.FilterTransactionsAll}`
    ),
    Markup.button.callback(
      `${formatMoney(row.income, bankAccount.currency)}`,
      `${BotAction.SelectStatisticsMonth}:${row.groupname}:${BotActionHidden.FilterTransactionsIncome}`
    ),
    Markup.button.callback(
      formatMoney(row.outcome, bankAccount.currency),
      `${BotAction.SelectStatisticsMonth}:${row.groupname}:${BotActionHidden.FilterTransactionsOutcome}`
    ),
  ];
};

export const buildMonthStatistics = (
  statisticRows: UserTransactionExpenseRowItem[],
  bankAccount: { id: string; currency: Currency }
) => {
  return [
    ...statisticRows.map((row) => buildStatisticGrid(row, bankAccount)),
    [
      Markup.button.callback(
        '◀️ Back',
        `${BotAction.SelectBankAccount}:${bankAccount.id}`
      ),
    ],
  ];
};

export const buildWeekStatistics = (
  statisticRows: UserTransactionExpenseRowItem[],
  bankAccount: BankAccount
) => {
  return [
    ...statisticRows.map((row) => buildStatisticGrid(row, bankAccount)),
    [
      Markup.button.callback(
        '◀️ Back',
        `${BotAction.SelectBankAccount}:${bankAccount.id}`
      ),
    ],
  ];
};

export const buildBankAccountListMenu = (bankAccounts: BankAccount[]) => {
  return [
    [
      Markup.button.callback(
        humanizeAction(BotAction.BankAccountAdd),
        BotAction.BankAccountAdd
      ),
    ],
    ...bankAccounts.map((bankAccount) => [
      Markup.button.callback(
        `${bankAccount.name} (${currencyToSymbol(bankAccount.currency)})`,
        `${BotAction.SelectBankAccount}:${bankAccount.id}`
      ),
    ]),
  ];
};
