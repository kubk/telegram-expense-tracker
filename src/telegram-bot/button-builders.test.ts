import { buildMonthStatistics } from './button-builders';
import { Currency } from '@prisma/client';

test('buildMonthStatistics', () => {
  expect(
    buildMonthStatistics(
      [
        {
          income: 1 * 100,
          currency: Currency.TRY,
          outcome: -10 * 100,
          groupname: '2',
          difference: -9 * 100,
        },
        {
          income: 1000 * 100,
          currency: Currency.TRY,
          outcome: -999 * 100,
          groupname: '1',
          difference: 1 * 100,
        },
      ],
      {
        id: '1',
        currency: 'TRY',
      }
    )
  ).toMatchSnapshot();
});
