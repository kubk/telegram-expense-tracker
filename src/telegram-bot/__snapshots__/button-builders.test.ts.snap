// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMonthStatistics 1`] = `
Array [
  Array [
    Object {
      "callback_data": "select_stats_month:2:all",
      "hide": false,
      "text": "2        -9₺",
    },
    Object {
      "callback_data": "select_stats_month:2:in",
      "hide": false,
      "text": "+1₺",
    },
    Object {
      "callback_data": "select_stats_month:2:out",
      "hide": false,
      "text": "-10₺",
    },
  ],
  Array [
    Object {
      "callback_data": "select_stats_month:1:all",
      "hide": false,
      "text": "1        +1₺",
    },
    Object {
      "callback_data": "select_stats_month:1:in",
      "hide": false,
      "text": "+1000₺",
    },
    Object {
      "callback_data": "select_stats_month:1:out",
      "hide": false,
      "text": "-999₺",
    },
  ],
  Array [
    Object {
      "callback_data": "select_bank_account:1",
      "hide": false,
      "text": "◀️ Back",
    },
  ],
]
`;
